presto-main/src/test/java/com/facebook/presto/operator/TestPartitionedOutputOperator.java

```diff
@@ -0,0 +1,311 @@
+import com.facebook.presto.operator.exchange.LocalPartitionGenerator;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+import com.facebook.presto.spi.block.RunLengthEncodedBlock;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.testing.TestingTaskContext;
+import com.facebook.presto.type.TypeRegistry;
+import com.google.common.collect.ImmutableList;
+import io.airlift.units.DataSize;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Random;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static com.facebook.presto.SessionTestUtils.TEST_SESSION;
+import static com.facebook.presto.block.BlockAssertions.createLongsBlock;
+import static com.facebook.presto.execution.buffer.BufferState.OPEN;
+import static com.facebook.presto.execution.buffer.BufferState.TERMINAL_BUFFER_STATES;
+import static com.facebook.presto.execution.buffer.OutputBuffers.BufferType.PARTITIONED;
+import static com.facebook.presto.execution.buffer.OutputBuffers.createInitialEmptyOutputBuffers;
+import static com.facebook.presto.memory.context.AggregatedMemoryContext.newSimpleAggregatedMemoryContext;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
+import static io.airlift.units.DataSize.Unit.BYTE;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+public class TestPartitionedOutputOperator
+{
+    private static final int PARTITION_COUNT = 512;
+    private static final DataSize MAX_MEMORY = new DataSize(1, GIGABYTE);
+    private static final DataSize PARTITION_MAX_MEMORY = new DataSize(5, MEGABYTE);
+    private static final List<Type> TYPES = ImmutableList.of(BIGINT);
+    private static final List<Type> REPLICATION_TYPES = ImmutableList.of(BIGINT, BIGINT);
+    private static final Random RANDOM = new Random(0);
+    private static final ExecutorService EXECUTOR = newCachedThreadPool(daemonThreadsNamed("test-EXECUTOR-%s"));
+    private static final ScheduledExecutorService SCHEDULER = newScheduledThreadPool(1, daemonThreadsNamed("test-%s"));
+    private static final int POSITION_COUNT = 1000;
+    private static final int NUM_OF_ITERATIONS = 100;
+
+    @Test
+    public void testOutputForSimplePage()
```

Comment:
Hardcoding expected values is problematic, because, it is hard for a reader to understand where these numbers come from and they need to be changed when parameters like POSITION_COUNT change.

I suggest to write these tests like so:

```
 private static final Block NULL_BLOCK = new RunLengthEncodedBlock(BIGINT.createBlockBuilder(null, 1).appendNull().build(), POSITION_COUNT);;
    private static final Block TESTING_BLOCK = createLongSequenceBlock(0, POSITION_COUNT);
    private static final Block TESTING_DICTIONARY_BLOCK = createLongDictionaryBlock(0, POSITION_COUNT);
    private static final Block TESTING_RLE_BLOCK = createRLEBlock(RANDOM.nextLong(), POSITION_COUNT);
    private static final Page TESTING_PAGE = new Page(TESTING_BLOCK);
    private static final Page TESTING_PAGE_WITH_NULL_BLOCK = new Page(POSITION_COUNT, NULL_BLOCK, createLongSequenceBlock(0, POSITION_COUNT));

    @Test
    public void testOutputForSimplePage()
    {
        PartitionedOutputOperator operator = createPartitionedOutputOperator(false);
        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
            operator.addInput(TESTING_PAGE);
        }
        operator.finish();

        OperatorContext context = operator.getOperatorContext();
        assertEquals(context.getOutputDataSize().getTotalCount(), NUM_OF_ITERATIONS * TESTING_PAGE.getSizeInBytes());
        assertEquals(context.getOutputPositions().getTotalCount(), NUM_OF_ITERATIONS * TESTING_PAGE.getPositionCount());
    }

    @Test
    public void testOutputForDictionaryPage()
    {
        PartitionedOutputOperator operator = createPartitionedOutputOperator(false);
        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
            operator.addInput(new Page(TESTING_DICTIONARY_BLOCK));
        }
        operator.finish();

        OperatorContext context = operator.getOperatorContext();
        assertEquals(context.getOutputDataSize().getTotalCount(), NUM_OF_ITERATIONS * TESTING_PAGE.getSizeInBytes());
        assertEquals(context.getOutputPositions().getTotalCount(), NUM_OF_ITERATIONS * TESTING_PAGE.getPositionCount());
    }

    @Test
    public void testOutputForRlePage()
    {
        PartitionedOutputOperator operator = createPartitionedOutputOperator(false);
        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
            operator.addInput(new Page(TESTING_RLE_BLOCK));
        }
        operator.finish();

        OperatorContext context = operator.getOperatorContext();
        assertEquals(context.getOutputDataSize().getTotalCount(), NUM_OF_ITERATIONS * TESTING_PAGE.getSizeInBytes());
        assertEquals(context.getOutputPositions().getTotalCount(), NUM_OF_ITERATIONS * TESTING_PAGE.getPositionCount());
    }

    @Test
    public void testOutputForSimplePageWithReplication()
    {
        PartitionedOutputOperator operator = createPartitionedOutputOperator(true);
        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
            operator.addInput(new Page(POSITION_COUNT, NULL_BLOCK, TESTING_BLOCK));
        }
        operator.finish();

        OperatorContext context = operator.getOperatorContext();
        assertEquals(context.getOutputDataSize().getTotalCount(), PARTITION_COUNT * TESTING_PAGE_WITH_NULL_BLOCK.getSizeInBytes() * NUM_OF_ITERATIONS);
        assertEquals(context.getOutputPositions().getTotalCount(), PARTITION_COUNT * TESTING_PAGE_WITH_NULL_BLOCK.getPositionCount() * NUM_OF_ITERATIONS);
    }

    @Test
    public void testOutputForDictionaryPageWithReplication()
    {
        PartitionedOutputOperator operator = createPartitionedOutputOperator(true);
        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
            operator.addInput(new Page(POSITION_COUNT, NULL_BLOCK, TESTING_DICTIONARY_BLOCK));
        }
        operator.finish();

        OperatorContext context = operator.getOperatorContext();
        assertEquals(context.getOutputDataSize().getTotalCount(),PARTITION_COUNT * TESTING_PAGE_WITH_NULL_BLOCK.getSizeInBytes() * NUM_OF_ITERATIONS);
        assertEquals(context.getOutputPositions().getTotalCount(),PARTITION_COUNT * TESTING_PAGE_WITH_NULL_BLOCK.getPositionCount() * NUM_OF_ITERATIONS);
    }

    @Test
    public void testOutputForRlePageWithReplication()
    {
        PartitionedOutputOperator operator = createPartitionedOutputOperator(true);
        for (int i = 0; i < NUM_OF_ITERATIONS; i++) {
            operator.addInput(new Page(POSITION_COUNT, NULL_BLOCK, TESTING_RLE_BLOCK));
        }
        operator.finish();

        OperatorContext context = operator.getOperatorContext();
        assertEquals(context.getOutputDataSize().getTotalCount(), PARTITION_COUNT * TESTING_PAGE_WITH_NULL_BLOCK.getSizeInBytes() * NUM_OF_ITERATIONS);
        assertEquals(context.getOutputPositions().getTotalCount(), PARTITION_COUNT * TESTING_PAGE_WITH_NULL_BLOCK.getPositionCount() * NUM_OF_ITERATIONS);
    }
```

Reply:
Thank you @mbasmanova . This is a great comment and helps in making test cases clear.